src/blog/2023/09/modernize-your-legacy-industrial-data-part2.md

```diff
@@ -0,0 +1,137 @@
+
+Let's take a look at a simple example. We have a process PLC recording production metrics and wish to get this data from its Modbus interface for reporting and decision making. The PLC has 5 values that we need to read:
+
+| Register | Value | Description            | Data Type  |
+| -------- | ----- | ---------------------- | ---------- |
+| 1, 2     | 10    | Part Count             | UINT32     |
+| 3, 4     | 2.5   | Cycle Time (sec)       | UINT32/100 |
+| 5, 6     | 30    | Production Time (sec)  | UINT32/100 |
+| 7, 8     | 20    | Run Time (sec)         | UINT32/100 |
+| 9, 10    | 11    | Stoppage Time (sec)    | UINT32/100 |
+
+If the total production time is 30 seconds and the run time is 20 seconds, then the stoppage time should be 10 seconds. But it is not. It is 11 seconds. This is because the data is not consistent.
+
+#### Why is the data inconsistent?
+
+The most common reason for the data inconsistency is that the data is being read from the PLC while the PLC is running. The data is changing as it is being read.
+
+Typically this happens when a developer, unfamiliar with the protocol or end device, begins by reading the data individually. Here is how this journey might look:
+
+_Image 1: individual reads_
+![image showing 10 individual reads](images/industrial-legacy-data-pt2-demo1.png)
+
+That is a lot of nodes and a lot of duplication!
+
+What is worse, is that the developer continues down this path and begins converting the data ready for publishing to MQTT. Here is how this might evolve:
+
+_Image 2: individual reads with data processing_
+![Image showing 10 individual reads with data processing](images/industrial-legacy-data-pt2-10polls.gif)
+
+Yippie! We have the data, it works, we publish it to MQTT, job done. Right?
+
+Unfortunately, no. The data is inconsistent.
+
+#### So whats the big deal?
+
+Inconsistent data is not useful data and errors can get compounded over time. This leads to bad decisions being made and a loss of confidence in the data. Ultimately, this leads to the data being ignored and the opportunities to make improvements are lost. Loss of improvements means loss of money!
+
+Not only that, from a developer or maintainers perspective, it has many problems:
+1. Lots of error prone manual configuration (yes, I made several mistakes while creating the example)
+2. Hard coded register addresses
+3. Duplication
+4. Inextensible - what if we need to read more registers?
+5. Inconsistent data - as discussed above
+6. Slow - each read takes time
+7. Inefficient use of network bandwidth - each read requires a request and response packet
+
+#### How can we make the data consistent?
+
+The most obvious solution is to stop the PLC before reading the data.
+This is of course a terrible idea as it will stop the process.
```

Comment:
I agree it's a terrible idea, but the phrasing is a bit aggressive for me.

Can we go more tong in cheek with

```suggestion
Though your boss won't like it as it will stop the manufacturing process.
```

Reply:
I am curious as to what makes you find that aggressive ZJ.  I normally recognise aggressive language and self correct (since I never mean to be aggressive) so I am genuinely curious as to the thoughts in your head here (for personal growth, nothing more 👼 )

I will of course update this but to let you into my head a little here: I described it as terrible (as we both agree it is) but not for a "terrible/bad-person" reason but a "terrible/strong-mistake" reason AKA - "really, don't do this!".  😃

Reply:
Applied change _similar_ to suggestion in [38834ba](https://github.com/flowforge/website/pull/1123/commits/38834bab3664f688aade56cdaebbe0c7175a7d9d)